Add a blog post about the interoperability gaps project (Q1 2023) (#72)
openwebdocs/owd-website: pages/content/posts/complete-interop-features/index.md

---
author: "Jean-Yves Perrier"
title: "Documenting missing interoperable web features"
date: "2023-04-17"
tags: ["projects"]
ShowToc: false
ShowBreadCrumbs: false
excerpt: "Filling some MDN gaps by documenting missing interoperable features."
---

> [MDN Web Docs](https://developer.mozilla.org/) is an invaluable set of information about the web platform. With about 12000 pages, it is the most extensive corpus of documents that web developers can refer to create their websites.
> 
> Unfortunately, despite being almost 20 years old, many reference pages still need to be included. We call these missing pages _gaps_ and even have tools to detect some of them:
> 
> - Mozilla's [Yari](https://github.com/mdn/yari/) renderer logs missing pages as flaws when it builds a page.
> - [Dominique Hazaël-Massieux](https://www.w3.org/People/Dom/) from the [W3C](https://www.w3.org/) created a tool ([mdn-gaps](https://dontcallmedom.github.io/mdn-gaps/)) listing all specification features without reference documentation. ![Extract of a screenshot listing a few random MDN gaps, as well as which browser supports them](mdn-gaps-example.png)
> 
> These tools allow us to know the extent of the problem.

Reviewer: ```suggestion
[MDN Web Docs](https://developer.mozilla.org/) is an invaluable information set about the web platform. With about 12000 pages, it is the most extensive corpus of documents that web developers can referene to create their websites.

Unfortunately, despite being almost 20 years old, many reference pages still need to be included. We call these missing pages _gaps_, and even have tools to detect some of them:

- Mozilla's [Yari](https://github.com/mdn/yari/) renderer logs missing pages as flaws when it builds a page.
- [Dominique Hazaël-Massieux](https://www.w3.org/People/Dom/), of the [W3C](https://www.w3.org/), created the ([mdn-gaps](https://dontcallmedom.github.io/mdn-gaps/)) tool which lists all specification features with missing reference documentation. 

![MDN gaps screen shot showing 5 missing fetch standard API pages and the browser support each feature](mdn-gaps-example.png)

These tools enable us to identify missing content and understand the extent of the issue.
```

Reviewer: the edits are minor (grammar). not sure why the darker green hiliting isn't kicking in.

## Why are there gaps on MDN?

> Gaps came organically. MDN Web Docs was, for many years, the home of a diverse set of documents: writers initially put much effort into documenting Firefox technologies like XUL and XBL, and time was scarce to focus on standardized languages. While CSS, HTML, and JavaScript had their documentation sets, docs for APIs were an afterthought. They were still living under a DOM hierarchy on MDN. With the advent of the HTML5 and CSS3 initiatives (and later the EcmaScript 6 one), Mozilla Developer Network's main subject shifted from proprietary to open and interoperable technologies. The shift was progressive, culminating in rebranding the site as _MDN Web Docs_, the archival of XUL and XBL docs, and, later, the move of Firefox Dev Tools docs to a dedicated [website](https://firefox-source-docs.mozilla.org/).

Reviewer: ```suggestion
Gaps came organically. MDN Web Docs is the home of a diverse set of documents. Initially, writers put great effort into documenting Firefox technologies like XUL and XBL, and time was scarce to focus on standardized languages. While CSS, HTML, and JavaScript were well documented, docs for APIs were an afterthought. They were still living under a DOM hierarchy on MDN. With the advent of HTML as a living standard and CSS module structure initiatives (and later the EcmaScript 6 one), Mozilla Developer Network's focus shifted from proprietary to open and interoperable technologies. The shift was progressive, culminating in rebranding the site as _MDN Web Docs_, the archiving of XUL and XBL docs, and, later, the move of [Firefox Dev Tools](https://firefox-source-docs.mozilla.org/) documentation to a dedicated website.
```

Reviewer: There are still diverse docs on MDN. 
I live to avoid HTML5 and CSS3 as deprecated terms.

date: "2023-04-17"
tags: ["projects"]
ShowToc: false
ShowBreadCrumbs: false
excerpt: "Filling some MDN gaps by documenting missing interoperable features."
---

[MDN Web Docs](https://developer.mozilla.org/) is an invaluable information set about the web platform. With about 12000 pages, it is the most extensive corpus of documents that web developers can referene to create their websites.

Unfortunately, despite being almost 20 years old, many reference pages still need to be included. We call these missing pages _gaps_, and even have tools to detect some of them:

- Mozilla's [Yari](https://github.com/mdn/yari/) renderer logs missing pages as flaws when it builds a page.
- [Dominique Hazaël-Massieux](https://www.w3.org/People/Dom/), of the [W3C](https://www.w3.org/), created the ([mdn-gaps](https://dontcallmedom.github.io/mdn-gaps/)) tool which lists all specification features with missing reference documentation. 

![MDN gaps screen shot showing 5 missing fetch standard API pages and the browser support each feature](mdn-gaps-example.png)

These tools enable us to identify missing content and understand the extent of the issue.

## Why are there gaps on MDN?

> Gaps came organically. MDN Web Docs is the home of a diverse set of documents. Initially, writers put great effort into documenting Firefox technologies like XUL and XBL, and time was scarce to focus on standardized languages. While CSS, HTML, and JavaScript were well documented, docs for APIs were an afterthought. They were still living under a DOM hierarchy on MDN. With the advent of HTML as a living standard and CSS module structure initiatives (and later the EcmaScript 6 one), Mozilla Developer Network's focus shifted from proprietary to open and interoperable technologies. The shift was progressive, culminating in rebranding the site as _MDN Web Docs_, the archiving of XUL and XBL docs, and, later, the move of [Firefox Dev Tools](https://firefox-source-docs.mozilla.org/) documentation to a dedicated website.

Reviewer 2: Consider omitting "Gaps came organically" and "MDN Web Docs is the home diverse documents" as the rest of the paragraph convey this connotation. 
```suggestion
Initially, the writers put great effort into documenting Firefox technologies like XUL and XBL, and time was scarce to focus on standardized languages. While CSS, HTML, and JavaScript were well documented, the docs for APIs were an afterthought. They were still living under a DOM hierarchy on MDN. With the advent of HTML as a living standard and CSS module structure initiatives (and later the EcmaScript 6 one), Mozilla Developer Network's focus shifted from proprietary to open and interoperable technologies. The shift was progressive, culminating in rebranding the site as _MDN Web Docs_, the archiving of XUL and XBL docs, and, later, the move of [Firefox Dev Tools](https://firefox-source-docs.mozilla.org/) documentation to a dedicated website.
```

Author: I agree the second sentence is not useful, but I would like to keep the first one: it is a kind of tl;dr:
```suggestion
Gaps came organically. Initially, the writers put great effort into documenting Firefox technologies like XUL and XBL, and time was scarce to focus on standardized languages. While CSS, HTML, and JavaScript were well documented, the docs for APIs were an afterthought. They were still living under a DOM hierarchy on MDN. With the advent of HTML as a living standard and CSS module structure initiatives (and later the EcmaScript 6 one), Mozilla Developer Network's focus shifted from proprietary to open and interoperable technologies. The shift was progressive, culminating in rebranding the site as _MDN Web Docs_, the archiving of XUL and XBL docs, and, later, the move of [Firefox Dev Tools](https://firefox-source-docs.mozilla.org/) documentation to a dedicated website.
```

> But, since the early 2010s, and even a bit earlier, standard bodies and browser makers rapidly specified and implemented new APIs. The focus was to document these new features: they were the focus of web developers, and there needed to be more good documentation about them! It was an easier move to sell internally: new technologies lead to more engagement; getting resources to document them was more straightforward than the old HTML table features, for example. The HTML5, CSS3, and ES6 initiatives allowed MDN to revamp these areas of documentation as there was an interest in positioning the web as a recognized interoperable platform.

Reviewer 2: I changed certain parts of the paragraph so that the history can be conveyed more clearly. I hope it helps! 😊
```suggestion
Since the early 2010s and previous eras, standard bodies and browser makers rapidly specified and implemented new APIs. The goal was to document these new features. Additionally, these features were the focus of web developers, thus, required more good documentation. It was an easier move to sell them internally as new technologies lead to more engagement. Additionally, getting resources to document these APIs were more straightforward than for, example, the old HTML table features. The HTML5, CSS3, and ES6 initiatives allowed MDN to revamp these areas of documentation because there was an interest in positioning the web as a recognized interoperable platform.
```

Author: ```suggestion
Since the early 2010s and in previous years, standard bodies and browser makers have rapidly specified and implemented new APIs. The goal was to document these new features. Additionally, these features were the focus of web developers, thus, required more good documentation. It was an easier move to sell them internally as new technologies led to more engagement. Additionally, getting resources to document these APIs was more straightforward than the old HTML table features. The HTML5, CSS3, and ES6 initiatives allowed MDN to revamp these areas of documentation because there was an interest in positioning the web as a recognized interoperable platform.

```

![The three logos of HTML5, CSS3, and JS](logos.png)

> So, the combination of the scarcity of resources and the high-speed development of the web platform led to the original gaps staying open. Some mitigating actions were taken, like listing every method and property for each interface, but there was still a large gap (more than 1000 missing pages).

Reviewer 2: ```suggestion
The combination of scarcity of resources and the high-speed development of the web platform led to the original gaps staying open. Some mitigating actions were taken, like listing every method and property for each interface. However,there was still a large gap (more than 1000 missing pages).
```

Today's gaps mostly fall into three categories:

- HTML DOM and the SVG DOM features added to the web platform when MDN was still in its infancy;
- non-interoperable features: APIs supported by only one or two browser engines and missed by the browser makers' writers;
- single properties and methods forgotten when their API was initially documented or added later (but they slipped below the radar of the different writing processes).

## Opportunities created by the Interop initiative

> As a collective of writers dedicated to helping document the open web platform, [Open Web Docs (OWD)](https://openwebdocs.org/) can help in _closing some of these gaps_. Unlike writers paid by browser makers needing to justify their work, our backers, individuals or companies, gave us the more generic mission to improve the open web documentation. We focus on much-needed activities that do not necessarily bring the most engagement at first, though improving in the middle term situation significantly. We still need to focus on the most valuable activities, but anybody's internal projects do not dictate our priorities.

Reviewer 2: ```suggestion
As a collective of writers dedicated to helping document the open web platform, [Open Web Docs (OWD)](https://openwebdocs.org/) can help in _closing some of these gaps_. Unlike writers who are paid by browser makers needing to justify their work, our backers, individuals or companies, gave us the more generic mission to improve the open web documentation. We focus on much-needed activities that do not necessarily bring the most engagement at first, though improving in the middle term situation significantly. We still need to focus on the most valuable activities, but anybody's internal projects do not dictate our priorities.
```

For several years now, browser makers have sat together at the end of each year to discuss the pain points of web developers and define a set of features that needs more love to increase their functionality or to fix bugs and incoherence so that web devs can use them reliably. They define areas to focus on as part of the [Interop](https://web.dev/interop-2023/) initiative.

> OWD, and its partners, think these projects are good opportunities to revisit the relevant area of the MDN Web Docs documentation: the idea is to have the docs in good shape for each set of features made interoperable over the year. Most of the time, this leads to minor changes to the docs: MDN writers tried to describe the interoperable behaviour, only mentioning browsers when their implementations diverged.

Reviewer 2: ```suggestion
OWD and its partners think these projects are good opportunities to revisit the relevant area of the MDN Web Docs documentation. The idea is to have the docs in good shape for each set of features made interoperable over the year. Most of the time, this leads to minor changes to the docs. MDN writers tried to describe the interoperable behavior, such as only mentioning browsers when their implementations diverged.
```

Author: I prefer the English spelling. (We are not on MDN where we use the American spelling)
```suggestion
OWD, and its partners, think these projects are good opportunities to revisit the relevant area of the MDN Web Docs documentation. The idea is to have the docs in good shape for each set of features made interoperable over the year. Most of the time, this leads to minor changes to the docs. MDN writers tried to describe the interoperable behaviour, such as only mentioning browsers when their implementations diverged.
```

> One of the Interop focus area this year, [interop-2023-forms](https://wpt.fyi/results/?label=master&label=experimental&product=chrome&product=firefox&product=safari&aligned&view=interop&q=label%3Ainterop-2022-forms%20or%20label%3Ainterop-2023-forms), is about making HTML Forms genuinely interoperable. It was an excellent opportunity to fill part of the MDN gaps. So we decided to do the following:

Reviewer 2: ```suggestion
One of the Interop focus area this year, [interop-2023-forms](https://wpt.fyi/results/?label=master&label=experimental&product=chrome&product=firefox&product=safari&aligned&view=interop&q=label%3Ainterop-2022-forms%20or%20label%3Ainterop-2023-forms), is about making HTML forms interoperable. It was an great opportunity to fill part of the MDN gaps. We decided to do the following:
```

1. Document all interoperable APIs unrelated to HTML and SVG.
2. Document all interoperable HTML APIs not part of the DOM (that is, everything in the HTML spec that is not an `HTML*Element` property or method).
3. Finally, document all missing HTML*Element features related to `<form>` (For example, `HTMLSelectElement`, `HTMLInputElement`, or `HTMLButtonElement`properties and methods).

> The two first points let us detect any newly interoperable property or method and document it if nobody has done it before. We already found some after the launch of Safari 16.4 (like [`WritableStreamDefaultController.signal`](https://developer.mozilla.org/en-US/docs/Web/API/WritableStreamDefaultController/signal)). These two initial steps will enable us to maintain the completeness of our API reference for interoperable features once we finish documenting the backlog.

Reviewer 2: What about the third step?

Author: ```suggestion
The two first points let us detect any newly interoperable property or method and document it if this is the first time anyone has done it. We already found some after the launch of Safari 16.4 (like [`WritableStreamDefaultController.signal`](https://developer.mozilla.org/en-US/docs/Web/API/WritableStreamDefaultController/signal)). These two initial steps will enable us to maintain the completeness of our API reference for interoperable features once we finish documenting the backlog.

The third step will give us a complete reference for the Forms APIs. It will be synchronized with this year's Interop initiative, where all three engines will be interoperable.
```

Already we can detect missed interoperable, unrelated to HTML and SVG DOMs, features that become interactive: for example, we added the [`CSSFontFeaturesValuesRule`](https://developer.mozilla.org/en-US/docs/Web/API/CSSFontFeatureValuesRule) that became interoperable in Q1.

In this Quarter, we will document HTML APIs not part of the HTML DOM, interfaces like `ErrorEvent`, `HTMLAllCollection`, or even `HTMLOptionsCollection`. You can follow our progress in our [tracking GitHub issue](https://github.com/openwebdocs/project/issues/153) In the future, writers will link numerous times to these documents. New pages will not have _red links_, links to non-written pages, improving the quality of our new docs. We also plan to improve the documentation for the _Constraint Validation API_ this Quarter.

Then, in Q3, we plan to tackle the last point, the missing `HTML*Element` features related to forms. Overall we plan to add 350 pages this year, about 30% of the overall gaps (for interoperable features).

## And all the rest?

> About 700 more interoperable properties and methods will not be documented after this first project. Even if we do not have a plan, we have a few ideas to tackle the remaining gaps. For SVG, we would like to take a similar approach: first, we would like to document fundamental interfaces, like the types: `SVGLength`, `SVGAnimatedLength`, … Then, in the second phase, we would like to call the community for help: there are a lot of similar docs to write, like the different kinds of SVG filters. The pages will have the same structure, although the examples will differ. We will see what will happen, and hopefully, we can report back on progress here!

Reviewer 2: ```suggestion
About 700 more interoperable properties and methods will not be documented after this first project. Even if we do not have a plan, we have a few ideas to tackle the remaining gaps. For SVG, we would like to take a similar approach. First, we would like to document fundamental interfaces, like the types: `SVGLength`, `SVGAnimatedLength`, … In the second phase, we would like to call the community for help. There are a lot of similar docs to write, like the different kinds of SVG filters. The pages will have the same structure, although the examples will differ. We will see what will happen, and hopefully, we can report back on progress here!
```

OWD, and its partners, think these projects are good opportunities to revisit the relevant area of the MDN Web Docs documentation. The idea is to have the docs in good shape for each set of features made interoperable over the year. Most of the time, this leads to minor changes to the docs. MDN writers tried to describe the interoperable behaviour, such as only mentioning browsers when their implementations diverged.

One of the Interop focus area this year, [interop-2023-forms](https://wpt.fyi/results/?label=master&label=experimental&product=chrome&product=firefox&product=safari&aligned&view=interop&q=label%3Ainterop-2022-forms%20or%20label%3Ainterop-2023-forms), is about making HTML forms interoperable. It was an great opportunity to fill part of the MDN gaps. We decided to do the following:

1. Document all interoperable APIs unrelated to HTML and SVG.
2. Document all interoperable HTML APIs not part of the DOM (that is, everything in the HTML spec that is not an `HTML*Element` property or method).

> 3. Finally, document all missing HTML*Element features related to `<form>` (For example, `HTMLSelectElement`, `HTMLInputElement`, or `HTMLButtonElement`properties and methods).

Author: ```suggestion
3. Finally, document all missing HTML*Element features related to `<form>` (For example, `HTMLSelectElement`, `HTMLInputElement`, or `HTMLButtonElement` properties and methods).
```

For several years now, browser makers have sat together at the end of each year to discuss the pain points of web developers and define a set of features that needs more love to increase their functionality or to fix bugs and incoherence so that web devs can use them reliably. They define areas to focus on as part of the [Interop](https://web.dev/interop-2023/) initiative.

OWD, and its partners, think these projects are good opportunities to revisit the relevant area of the MDN Web Docs documentation. The idea is to have the docs in good shape for each set of features made interoperable over the year. Most of the time, this leads to minor changes to the docs. MDN writers tried to describe the interoperable behaviour, such as only mentioning browsers when their implementations diverged.

> One of the Interop focus area this year, [interop-2023-forms](https://wpt.fyi/results/?label=master&label=experimental&product=chrome&product=firefox&product=safari&aligned&view=interop&q=label%3Ainterop-2022-forms%20or%20label%3Ainterop-2023-forms), is about making HTML forms interoperable. It was an great opportunity to fill part of the MDN gaps. We decided to do the following:

Reviewer: ```suggestion
One of the Interop focus areas this year, [interop-2023-forms](https://wpt.fyi/results/?label=master&label=experimental&product=chrome&product=firefox&product=safari&aligned&view=interop&q=label%3Ainterop-2022-forms%20or%20label%3Ainterop-2023-forms), is to make HTML forms interoperable. It is a great opportunity to fill some of MDN's gaps. We decided to do the following:
```

1. Document all interoperable APIs unrelated to HTML and SVG.
2. Document all interoperable HTML APIs not part of the DOM (that is, everything in the HTML spec that is not an `HTML*Element` property or method).

> 3. Finally, document all missing HTML*Element features related to `<form>` (For example, `HTMLSelectElement`, `HTMLInputElement`, or `HTMLButtonElement` properties and methods).

Reviewer: ```suggestion
3. Finally, document all missing HTML*Element features related to `<form>` (For example, `HTMLSelectElement`, `HTMLInputElement`, and `HTMLButtonElement` properties and methods).
```

> The two first points let us detect any newly interoperable property or method and document it if this is the first time anyone has done it. We already found some after the launch of Safari 16.4 (like [`WritableStreamDefaultController.signal`](https://developer.mozilla.org/en-US/docs/Web/API/WritableStreamDefaultController/signal)). These two initial steps will enable us to maintain the completeness of our API reference for interoperable features once we finish documenting the backlog.

Reviewer: ```suggestion
The two first points let us detect any newly interoperable property or method and document it if no one else has already done so. For example, we found gaps after the launch of Safari 16.4 (like [`WritableStreamDefaultController.signal`](https://developer.mozilla.org/en-US/docs/Web/API/WritableStreamDefaultController/signal)). These two initial steps will enable us to maintain the completeness of our API reference for interoperable features (once we finish documenting the backlog).
```

> The third step will give us a complete reference for the Forms APIs. It will be synchronized with this year's Interop initiative, where all three engines will be interoperable.

Reviewer: ```suggestion
The third step will give us a complete reference for the Forms APIs. It will be synchronized with this year's Interop initiative, where all three engines will be interoperable, fully supporting the Forms APIs.
```

Reviewer: my edit may be redundant, but i just wanted to remind readers what interop means.

## Current progress

> During the first Quarter of 2023, we focused on the first point, documenting all interoperable APIs unrelated to HTML and SVG (See our [tracking project](https://github.com/openwebdocs/project/issues/152)). We added more than 100 pages to MDN, documenting missing modern features like [`CSSLayerBlockRule`](https://developer.mozilla.org/en-US/docs/Web/API/CSSLayerBlockRule) and [`CSSLayerStatementRule`](https://developer.mozilla.org/en-US/docs/Web/API/CSSLayerStatementRule), or old features writers forgot over the years, like [`XMLHttpRequestUpload`](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequestUpload).

Reviewer: ```suggestion
During the first Quarter of 2023, we focused on the first point, documenting all interoperable APIs outside of the HTML and SVG APIs (See our [tracking project](https://github.com/openwebdocs/project/issues/152)). We added more than 100 pages to MDN, documenting missing modern features like [`CSSLayerBlockRule`](https://developer.mozilla.org/en-US/docs/Web/API/CSSLayerBlockRule) and [`CSSLayerStatementRule`](https://developer.mozilla.org/en-US/docs/Web/API/CSSLayerStatementRule), or old features writers forgot over the years, like [`XMLHttpRequestUpload`](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequestUpload).
```

> Already we can detect missed interoperable, unrelated to HTML and SVG DOMs, features that become interactive: for example, we added the [`CSSFontFeaturesValuesRule`](https://developer.mozilla.org/en-US/docs/Web/API/CSSFontFeatureValuesRule) that became interoperable in Q1.

Reviewer: ```suggestion
Already we can detect missed interoperable APIs, unrelated to HTML and SVG DOMs, features that become interoperative: for example, we added the [`CSSFontFeaturesValuesRule`](https://developer.mozilla.org/en-US/docs/Web/API/CSSFontFeatureValuesRule) that received support in all major browsers in Q1.
```

> In this Quarter, we will document HTML APIs not part of the HTML DOM, interfaces like `ErrorEvent`, `HTMLAllCollection`, or even `HTMLOptionsCollection`. You can follow our progress in our [tracking GitHub issue](https://github.com/openwebdocs/project/issues/153) In the future, writers will link numerous times to these documents. New pages will not have _red links_, links to non-written pages, improving the quality of our new docs. We also plan to improve the documentation for the _Constraint Validation API_ this Quarter.

Reviewer: ```suggestion
In Q2, we are documenting HTML APIs that are part of the HTML DOM, interfaces like `ErrorEvent`, `HTMLAllCollection`, or even `HTMLOptionsCollection`. You can follow our progress in our [GitHub tracking issue](https://github.com/openwebdocs/project/issues/153). In the future, writers will link to these documents. New pages will not have _red links_, links to non-existent pages, improving the quality of our new documentation. The _Constraint Validation API_ documentation will also be improved this Quarter.
```

Then, in Q3, we plan to tackle the last point, the missing `HTML*Element` features related to forms. Overall we plan to add 350 pages this year, about 30% of the overall gaps (for interoperable features).

## And all the rest?

> About 700 more interoperable properties and methods will not be documented after this first project. Even if we do not have a plan, we have a few ideas to tackle the remaining gaps. For SVG, we would like to take a similar approach. First, we would like to document fundamental interfaces, like the types: `SVGLength`, `SVGAnimatedLength`, … In the second phase, we would like to call the community for help. There are a lot of similar docs to write, like the different kinds of SVG filters. The pages will have the same structure, although the examples will differ. We will see what will happen, and hopefully, we can report back on progress here!

Reviewer: ```suggestion
After this interop documentation project is complete, almost 700 more interoperable properties and methods will still need to be documented. We are developing plans to tackle the remaining gaps. 

SVG documentation could benefit from a similar interop approach. In a first phase, we would like to document fundamental interfaces, like the types: `SVGLength`, `SVGAnimatedLength`. We would like to call the community for help in the second phase: there are a lot of similar docs to write, like the multitude of SVG filters. Pages with similar structures, though with different examples, are a great opportunity for community involvement. 

We will see what will happen, and hopefully, we can report back on progress here!
```

Reviewer: I didn't want to say "first" project, since this is the second interop (or third) as interop 2022 was a project, and interop 2023 is too, though that one hasn't started so not sure if that is 2nd or 3rd

---
author: "Jean-Yves Perrier"
title: "Documenting missing interoperable web features"

> date: "2023-04-17"

Reviewer: ```suggestion
date: "2023-04-25"
```

tags: ["projects"]
ShowToc: false
ShowBreadCrumbs: false
excerpt: "Filling some MDN gaps by documenting missing interoperable features."
---

[MDN Web Docs](https://developer.mozilla.org/) is an invaluable information set about the web platform. With about 12,000 pages, it is the most extensive corpus of documents that web developers can rely on when creating websites.

> Unfortunately, despite being almost 20 years old, many reference pages still need to be added. We call these missing pages _gaps_, and even have tools to detect some of them:

Reviewer: ```suggestion
Unfortunately, despite being almost 20 years old, many reference pages still need to be added. We call these missing pages _gaps_, and even have tools to identify missing content and understand the extent of the issue.:
```

- Mozilla's [Yari](https://github.com/mdn/yari/) renderer logs missing pages as flaws when it builds a page.

> - [Dominique Hazaël-Massieux](https://www.w3.org/People/Dom/), of the [W3C](https://www.w3.org/), created the ([mdn-gaps](https://dontcallmedom.github.io/mdn-gaps/)) tool which lists all specification features with missing reference documentation. 

Reviewer: ```suggestion
- [Dominique Hazaël-Massieux](https://www.w3.org/People/Dom/), of the [W3C](https://www.w3.org/), created the [mdn-gaps](https://dontcallmedom.github.io/mdn-gaps/) tool which lists all specification features with missing reference documentation. 
```

![MDN gaps screen shot showing 5 missing fetch standard API pages and the browser support each feature](mdn-gaps-example.png)

> These tools enable us to identify missing content and understand the extent of the issue.

Reviewer: ```suggestion
```
said this above

## Why are there gaps on MDN?

> Missing pages or gaps slipped in and their number grew organically. Initially, most MDN writers put great effort into documenting Firefox technologies like XUL and XBL, and time was scarce to focus on standardized languages. While CSS, HTML, and JavaScript were well documented, originally the docs for Web APIs were an afterthought. They lived under a DOM hierarchy on MDN. With the advent of HTML as a living standard and CSS module structure initiatives (and later ECMAScript 6), Mozilla Developer Network's focus shifted more from documenting proprietary to document open and interoperable technologies. The shift was progressive, culminating in rebranding the site as _MDN Web Docs_, the archiving of XUL and XBL docs, and, later, the moving of [Firefox Dev Tools](https://firefox-source-docs.mozilla.org/) documentation to a dedicated website.

Reviewer: ```suggestion
Missing pages or gaps slipped in and their number grew organically. Initially, most MDN writers put great effort into documenting Firefox technologies like XUL and XBL, and time was scarce to focus on standardized languages. While CSS, HTML, and JavaScript were well documented, originally the docs for Web APIs were an afterthought. They lived under a DOM hierarchy on MDN. With the advent of HTML as a living standard and CSS module structure initiatives (and later ECMAScript 6), Mozilla Developer Network's focus shifted more from documenting proprietary content to documenting open and interoperable technologies. The shift was progressive, culminating in rebranding the site as _MDN Web Docs_, the archiving of XUL and XBL docs, and, later, the moving of [Firefox Dev Tools](https://firefox-source-docs.mozilla.org/) documentation to a dedicated website.
```

Since the early 2010s (and even prior), standard bodies and browser makers have rapidly specified and implemented new APIs. MDN's goal was to document these new features. As these features were web developers' main focus, extensive, accurate, and current documentation was (and continues to be) required. The originally named "HTML5", "CSS3", and "ES6" initiatives allowed MDN to revamp these areas of documentation because there was an interest in positioning the web as a recognized interoperable and modern platform.

![The three logos of HTML5, CSS3, and JS](logos.png)

> For the MDN writing community, the development of the web platform at larger speed led to being unable to keep up with documenting all new features while also trying to document the existing foundation of the web platform. This sometimes lead to mitigating actions, like listing methods and properties on the interface page but not creating the sub pages.

Reviewer: ```suggestion
The development of the web platform at a greater speed led to the MDN writing community being unable to keep up with documenting all new features while also trying to document the existing foundations of the web platform. This sometimes led to mitigating actions, like listing methods and properties on the interface page but not creating the sub-pages.
```

Today's gaps mostly fall into three categories:

> - HTML DOM, and the SVG DOM features added to the web platform when MDN was still in its infancy;

Reviewer: ```suggestion
- HTML DOM and SVG DOM features added to the web platform when MDN was still in its infancy;
```

- non-interoperable features: APIs supported by only one or two browser engines;
- single properties and methods that were forgotten when the interface was initially documented or added later.

## Opportunities created by the Interop initiative

As a collective of writers dedicated to documenting the open web platform, [Open Web Docs (OWD)](https://openwebdocs.org/) is able to _close these gaps_.  We believe that projects like this can significantly improve in the middle- and long-term health and the longevity of the documentation. We aim to provide complete documentation of the web platform.

> For several years now, browser makers have come together to discuss the pain points of web developers and define the set of features that need more attention to improve their cross-browser functionality and to fix inconsistencies and issues ensuring web devs can implement them reliably. They define areas to focus on as part of the [Interop](https://web.dev/interop-2023/) initiative.

Reviewer: ```suggestion
For several years now, browser makers have come together to discuss web developer pain points, defining the features that need more attention to improve cross-browser functionality and fixing inconsistencies and issues, ensuring web devs can implement features reliably. This is known as the [Interop](https://web.dev/interop-2023/) initiative.
```

> As a collective of writers dedicated to helping document the open web platform, [Open Web Docs (OWD)](https://openwebdocs.org/) is helping _close these gaps_. Unlike writers who are paid by browser makers needing to justify their work, OWD's backers, individuals and companies, gave us the more generic mission to improve open web documentation. We are able to focus on much-needed activities that may not bring the most engagement at first, but significantly improve in the middle- and long-term situations. We are able to focus on these valuable activities that benefit developers and the longevity of the documentation, without having anybody's internal projects dictate our priorities.

Reviewer 3: ```suggestion
As a collective of writers dedicated to documenting the open web platform, [Open Web Docs (OWD)](https://openwebdocs.org/) is able to _close these gaps_.  We believe that projects like this can significantly improve in the middle- and long-term health and the longevity of the documentation. We aim to provide complete documentation of the web platform.
```

> OWD's Steering Committee has determined these projects provide excellent opportunities to revisit the relevant areas of MDN Web Docs documentation. The objective and goal is to have complete documentation for each set of features made interoperable over the year. 

Reviewer 4: nit: maybe link to the steering committe page on OWD's website.

Author: Great idea!

> Within OWD's Steering Committee, we have determined that these projects are good opportunities to revisit the relevant area of the MDN Web Docs documentation. The idea is to have the docs in good shape for each set of features made interoperable over the year. Most of the time, this leads to minor changes to the docs. MDN writers tried to describe the interoperable behaviour, such as only mentioning browsers when their implementations diverged.

Reviewer: ```suggestion
OWD's Steering Committee has determined these projects provide excellent opportunities to revisit the relevant areas of MDN Web Docs documentation. The objective and goal is to have complete documentation for each set of features made interoperable over the year. 

Most of the time, this leads to minor changes to the documentation. MDN writers have described the interoperable behavior, such as mentioning when and how browser implementations have diverged; with interoperability, notes about divergence in browser behaviors can be removed.
```

> OWD, and its partners, think these projects are good opportunities to revisit the relevant area of the MDN Web Docs documentation. The idea is to have the docs in good shape for each set of features made interoperable over the year. Most of the time, this leads to minor changes to the docs. MDN writers tried to describe the interoperable behaviour, such as only mentioning browsers when their implementations diverged.

Reviewer 3: ```suggestion
Within OWD's Steering Committee, we have determined that these projects are good opportunities to revisit the relevant area of the MDN Web Docs documentation. The idea is to have the docs in good shape for each set of features made interoperable over the year. Most of the time, this leads to minor changes to the docs. MDN writers tried to describe the interoperable behaviour, such as only mentioning browsers when their implementations diverged.
```

One of the Interop focus areas this year, [interop-2023-forms](https://wpt.fyi/results/?label=master&label=experimental&product=chrome&product=firefox&product=safari&aligned&view=interop&q=label%3Ainterop-2022-forms%20or%20label%3Ainterop-2023-forms), is to make HTML forms interoperable. It is a great opportunity to fill some of MDN's gaps. We decided to do the following:

1. Document all interoperable APIs unrelated to HTML and SVG.
2. Document all interoperable HTML APIs not part of the DOM (that is, everything in the HTML spec that is not an `HTML*Element` property or method).

> 3. Finally, document all missing HTML*Element features related to `<form>` (For example, `HTMLSelectElement`, `HTMLInputElement`, and `HTMLButtonElement` properties and methods).

Reviewer 4: ```suggestion
3. Finally, document all missing `HTML*Element` features related to `<form>` (For example, `HTMLSelectElement`, `HTMLInputElement`, and `HTMLButtonElement` properties and methods).
```

The third step will give us a complete reference for the Forms APIs. It will be synchronized with this year's Interop initiative, where all three engines will be interoperable, fully supporting the Forms APIs.

## Current progress

During the first Quarter of 2023, we focused on the first point, documenting all interoperable APIs outside of the HTML and SVG APIs (See our [tracking project](https://github.com/openwebdocs/project/issues/152)). We added more than 100 pages to MDN, documenting missing modern features like [`CSSLayerBlockRule`](https://developer.mozilla.org/en-US/docs/Web/API/CSSLayerBlockRule) and [`CSSLayerStatementRule`](https://developer.mozilla.org/en-US/docs/Web/API/CSSLayerStatementRule), or old features writers forgot over the years, like [`XMLHttpRequestUpload`](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequestUpload).

> Already we can detect missed interoperable APIs, unrelated to HTML and SVG DOMs, features that become interoperative: for example, we added the [`CSSFontFeaturesValuesRule`](https://developer.mozilla.org/en-US/docs/Web/API/CSSFontFeatureValuesRule) that received support in all major browsers in Q1.

Reviewer 4: ```suggestion
Already we can detect missed interoperable APIs, unrelated to HTML and SVG DOMs, features that become interoperative: for example, we added the [`CSSFontFeaturesValuesRule`](https://developer.mozilla.org/en-US/docs/Web/API/CSSFontFeatureValuesRule) that received support in all major browsers in the first quarter of 2023.
```

I think it's clearer this way, Q1 might be a bit cryptic for some people, and the year is helpful because this blog post will live on.
However, I'm not sure if you meant that support was added in Q1, or if we added documentation in Q1. If the latter, then it might make sense to move the Q1 bit earlier in the sentence, closer to "we added".

> In Q2, we are documenting HTML APIs that are part of the HTML DOM, interfaces like `ErrorEvent`, `HTMLAllCollection`, or even `HTMLOptionsCollection`. You can follow our progress in our [GitHub tracking issue](https://github.com/openwebdocs/project/issues/153). In the future, writers will link to these documents. New pages will not have _red links_, links to non-existent pages, improving the quality of our new documentation. The _Constraint Validation API_ documentation will also be improved this Quarter.

Reviewer 4: ```suggestion
In the second quarter, we are documenting HTML APIs that are part of the HTML DOM, interfaces like `ErrorEvent`, `HTMLAllCollection`, or even `HTMLOptionsCollection`. You can follow our progress in our [GitHub tracking issue](https://github.com/openwebdocs/project/issues/153). In the future, writers will link to these documents. New pages will not have _red links_, links to non-existent pages, improving the quality of our new documentation. The _Constraint Validation API_ documentation will also be improved this Quarter.
```

> Then, in Q3, we plan to tackle the last point, the missing `HTML*Element` features related to forms. Overall we plan to add 350 pages this year, about 30% of the overall gaps (for interoperable features).

Reviewer 4: ```suggestion
Later this year, we plan to tackle the last point, the missing `HTML*Element` features related to forms. Overall we plan to add 350 pages this year, about 30% of the overall gaps (for interoperable features).
```

## And all the rest?

After this interop documentation project is complete, almost 700 more interoperable properties and methods will still need to be documented. We are developing plans to tackle the remaining gaps.

> SVG documentation could benefit from a similar interop approach. In the first phase, we would like to document fundamental interfaces, like the types: `SVGLength`, `SVGAnimatedLength`. We want to call the community for help in the second phase: there are a lot of similar docs to write, like the multitude of SVG filters. Pages with similar structures, though with different examples, are a great opportunity for community involvement. 

Reviewer 4: The last part of this paragraph is an awesome call to action, unfortunately there's no link for someone to actually get involved. Can we link to a set of issues related to this? Or to the contribution guidelines?

Author: I need to create an MDN discussion about completing this. The problem is that we must first complete documentation of SVG types (like SVGLength and SVGAnimatedLength). 

I will try to create this discussion before publishing this. If not, I'll add the link later.

---
author: "Jean-Yves Perrier"
title: "Documenting missing interoperable web features"
date: "2023-04-25"
tags: ["projects"]
ShowToc: false
ShowBreadCrumbs: false
excerpt: "Filling some MDN gaps by documenting missing interoperable features."
---

[MDN Web Docs](https://developer.mozilla.org/) is an invaluable information set about the web platform. With about 12,000 pages, it is the most extensive corpus of documents that web developers can rely on when creating websites.

Unfortunately, despite being almost 20 years old, many reference pages still need to be added. We call these missing pages _gaps_, and even have tools to identify missing content and understand the extent of the issue:

- Mozilla's [Yari](https://github.com/mdn/yari/) renderer logs missing pages as flaws when it builds a page.
- [Dominique Hazaël-Massieux](https://www.w3.org/People/Dom/), of the [W3C](https://www.w3.org/), created the [mdn-gaps](https://dontcallmedom.github.io/mdn-gaps/) tool which lists all specification features with missing reference documentation. 

![MDN gaps screen shot showing 5 missing fetch standard API pages and the browser support each feature](mdn-gaps-example.png)

> 

Author: ```suggestion
```
